Prepare for cross-compilability by using real_t as compile-time real type

kinke opened the pull request:
And by removing the `longdouble` alias.

Also extract the `real_t` stuff from `ddmd.root.port` into its own module, `ddmd.root.real_t`. Cross-compilers (GDC, LDC in the near future) should only have to re-implement that module, including a custom `real_t` type implementing any desired target `real` precision in software.

Also includes some clean-ups, refactorings and renamings.

kinke committed:
Prepare for cross-compilability by using real_t as target real type

And by removing the `longdouble` alias.

Also extract the real_t stuff from ddmd.root.port into its own module,
ddmd.root.real_t. Cross-compilers (GDC, LDC in the near future) should
only have to re-implement that module, including a custom real_t type
implementing any desired target `real` precision in software.

Also includes some clean-ups, refactorings and renamings.

ibuclaw commented:
As per PR in ldc...

kinke commented:
Nice and unexpected side effect: this reveals a [mangling bug](https://github.com/ldc-developers/ldc/issues/1312) in LDC. ;)

ibuclaw commented on src/builtin.d:
```diff
@@ -69,7 +69,7 @@ extern (C++) Expression eval_sqrt(Loc loc, FuncDeclaration fd, Expressions* argu
     Expression arg0 = (*arguments)[0];
     assert(arg0.op == TOKfloat64);
-    return new RealExp(loc, Port.sqrt(arg0.toReal()), arg0.type);
+    return new RealExp(loc, ddmd.root.real_t.sqrt(arg0.toReal()), arg0.type);
```
This, and other functions in ``Port`` should remain there, as they have nothing to do with ``real_t`` per say, other than they accept and return that type.

kinke commented:
@ibuclaw: I hope you haven't started working on your rebase yet. ;) I tried to incorporate all your suggestions.

ibuclaw commented:
@kinke - you are too impatient. :-)

kinke commented:
@ibuclaw: How about only moving the `yl2x` and `isFloat<XX>LiteralOutOfRange()` back into the `Port` struct?
The rest (`snan` global, `sqrt()`, `areBitwiseEqual()`, `isNaN()`, `isSignallingNaN()`, `isInfinity()`, `parseTargetReal()` and `sprint()`) seems ideal in `real_t.d` to me...

kinke commented on src/root/port.d:
```diff
@@ -52,40 +42,21 @@ extern (C++) struct Port
+            assert(0);
     }
 
-    static real fequal(real a, real b)
```
A `real` should be big enough to store the bool. I went with a `bool` nevertheless. ;)

ibuclaw commented:
I'd be happy leaving everything in ``Port``, and just having things forwarded to ``real_t`` from there if needed.

The less the front-end knows about ``real_t``, the better.

kinke commented on src/lexer.d:
```diff
@@ -2144,7 +2138,8 @@ public:
         }
-        if (errno == ERANGE)
+        const isLong = (result == TOKfloat80v || result == TOKimaginary80v);
+        if (isOutOfRange && !isLong)
```
This prepares a bugfix wrt. this ignoring of out-of-range real literals. I can't just error out, as disabled code blocks are still lexed - e.g., unit tests specifically only enabled for x87 reals with huge literals.

kinke commented:
With these changes and the consistent usage of `real_t` (there's only a single occurrence of `real` in the front-end left - as `real_t` alias), the `real_t` type is quite wide-spread (as opposed to the `Port` struct), and so are the imports in other modules. I think the closely coupled utility functions belong exclusively there in a more emancipated `ddmd.root.real_t` module, emphasizing its key role for cross-compilation.

ibuclaw commented on src/root/real_t.d:
```diff
@@ -0,0 +1,118 @@
+    {
+        import core.math;
+        return core.math.sqrt(x);
+    }
```
This should remain in Port too.

ibuclaw commented on src/constfold.d:
```diff
@@ -450,12 +449,12 @@ extern (C++) UnionExp Mod(Loc loc, Type type, Expression e1, Expression e2)
         {
             real_t r2 = e2.toReal();
-            c = complex_t(Port.fmodl(e1.toReal(), r2), Port.fmodl(e1.toImaginary(), r2));
+            c = complex_t(e1.toReal() % r2, e1.toImaginary() % r2);
```
Eek!?!

kinke replied:
What? Let's require `real_t` to support the modulo operator instead of going through `Port`...

ibuclaw replied:
Yes, I looked at it first thing in the morning, and didn't consider operator overloading at that time.

ibuclaw commented on src/root/real_t.d:
```diff
@@ -0,0 +1,118 @@
+        return core.math.sqrt(x);
+    }
+
+    bool areBitwiseEqual(real_t a, real_t b)
```
isIdentical

ibuclaw commented on src/root/real_t.d:
```diff
@@ -0,0 +1,118 @@
+        return r;
+    }
+
+    int printTargetReal(char* str, char fmt, real_t x)
```
This should be 'sprintXXX' to suggest that output is written to buffer, and not stdout.

WalterBright commented on src/builtin.d:
```diff
@@ -150,10 +151,10 @@ extern (C++) Expression eval_yl2x(Loc loc, FuncDeclaration fd, Expressions* argu
-    real y = arg1.toReal();
-    real result;
-    Port.yl2x_impl(&x, &y, &result);
+    real_t x = arg0.toReal();
```
Global comment: use `auto` or (even better) `const` where possible instead of `real_t`.

WalterBright commented on src/dmangle.d:
```diff
@@ -630,17 +629,17 @@ public:
          * 0X1.9P+2                 => 19P2
          */
-        if (Port.isNan(value))
+        if (isNaN(value))
```
Where is it getting `isNaN` from? Since there are many `isNan` functions, and `isInfinity` functions, I'd prefer to see this name qualified.

kinke replied:
No problem - would you prefer the fully qualified name or putting these functions as static functions into a helper struct (analogous to `Port`)?

yebblies replied:
Why cares, they all do the same thing?

kinke replied:
Okay, a `TargetReal` struct it shall be then. ;)

yebblies replied:
Sorry, that was addressed to Walter.  I'd say it goes with TargetReal if it depends on the real implementation.

WalterBright replied:
> Why cares, they all do the same thing?

One would hope so, but I've seen enough bad implementations of those isXXXX floating point functions that I don't trust pulling some random one from somewhere. That goes for all the math functions, too. Turns out that C library programmers tend to stink at floating point.

I once attended a talk by a physics prof on doing numerical work. He was astonished when I told him the C library sin(), cos(), etc., could not be trusted. He didn't believe me. I know from bitter experience. The usual faults they have are truncated precision and erratic handling of special values.

So, in code I need to trust, I need to see **which** implementation is being called.

WalterBright commented on src/dmangle.d:
```diff
@@ -630,17 +629,17 @@ public:
+            enum BUFFER_LEN = 36;
             char[BUFFER_LEN] buffer;
-            size_t n = Port.ld_sprint(buffer.ptr, 'A', value);
+            int n = printTargetReal(buffer.ptr, 'A', value);
```
Replace `int` with `const`

WalterBright commented on src/expression.d:
```diff
@@ -2009,7 +2008,7 @@ private:
 extern (C++) int RealEquals(real_t x1, real_t x2)
 {
-    return (Port.isNan(x1) && Port.isNan(x2)) || Port.fequal(x1, x2);
+    return (isNaN(x1) && isNaN(x2)) || areBitwiseEqual(x1, x2);
```
qualify `isNaN`

WalterBright commented on src/target.d:
```diff
@@ -348,23 +348,23 @@ extern (C++) static void encodeReal(Expression e, ubyte* buffer)
 }
 
-// Write the bytes encoded in 'buffer' into a longdouble and returns
+// Write the bytes encoded in 'buffer' into a real and returns
```
`real_t`

kinke commented:
Reworked to incorporate your suggestions. I stumbled upon additional built-ins (sin/cos/tan/fabs besides sqrt) and hope to have them all now. ;)

ibuclaw commented on src/expression.d:
```diff
@@ -2009,7 +2008,7 @@ private:
 extern (C++) int RealEquals(real_t x1, real_t x2)
 {
-    return (Port.isNan(x1) && Port.isNan(x2)) || Port.fequal(x1, x2);
+    return (TargetReal.isNaN(x1) && TargetReal.isNaN(x2)) || TargetReal.areIdentical(x1, x2);
```
``areIdentical``?  I think it's best to stick with ``std.math`` name conventions. :-)

kinke replied:
Will fix - I haven't used `std.math` in ages. ;)

ibuclaw commented on src/builtin.d:
```diff
@@ -48,35 +49,35 @@ extern (C++) Expression eval_sin(Loc loc, FuncDeclaration fd, Expressions* argum
     Expression arg0 = (*arguments)[0];
     assert(arg0.op == TOKfloat64);
-    return new RealExp(loc, sinl(arg0.toReal()), arg0.type);
+    return new RealExp(loc, TargetReal.sin(arg0.toReal()), arg0.type);
```
Again, I'm not sure what these math functions have to do with the layout/properties of ``real_t``.  I really do think they should all be part of ``Port``, which is already used for doing things on the host in an agnostic manner (even though DMD may not follow that precise specification, it's how GDC treats it).

kinke replied:
And once again, that's where we disagree. I see no advantage by moving these helpers back to `Port`, see my arguments before. They **only** depend on the `real_t` type and don't just take and return one (they should actually be considered `pure`).

ibuclaw replied:
And what advantage are there leaving them out?

kinke replied:
Simple:

* Decoupling both modules: with a custom `real_t` type, one only has to touch 1 module, leading to either
  * no redundant forwarders from `Port` to the actual implementations in `real_t.d`, or
  * no `real_t` specific implementations in `port.d` (excluding `yl2x` for a moment).
* Less imports in other modules.

ibuclaw replied:
But ``port.d`` is not shareable anyway.  You can cut the forwarding in ``Port`` by just inlining it in LDC's implementation.

kinke replied:
Huh? We are talking about the interface required by the front-end (`core.builtin` etc.) here, which I prefer centralized in `real_t.d` and you'd prefer split between `real_t.d` (`real_t` type only) and `port.d`.
The new `port.d` is pretty short, and I'm not sure we'll need to adapt it for LDC (possibly for big-endian platforms).

I'm already working on a `real_t.d` implementation for LDC, based on @redstar's work around LLVM's `APFloat` type. That type doesn't seem to offer sin/cos/tan/sqrt, so that's going to be an issue - are these required built-ins available for the type you use in GDC?

ibuclaw replied:
>>  are these required built-ins available for the type you use in GDC?

Of course they aren't!  Built-ins aren't handled explicitly in GDC, instead a tree is generated and passed to const-fold in the middle-end to see if it returns a literal.

ibuclaw replied:
real_t also concerns the host platform, not the target.

WalterBright replied:
Port's reason for existence is to remove #if stuff from the front end.

ibuclaw replied:
> Port's reason for existence is to remove #if stuff from the front end.

It also implements functionality that may not be available on every platform (``memicmp``/``stricmp``).  I think it is safe to say that long double math functions also falls into this category too, it's just that DMD has never had enough widespread support to run into the problem.

ibuclaw replied:
> That type doesn't seem to offer sin/cos/tan/sqrt, so that's going to be an issue

By the way, these functions are implemented by the MPFR library, which requires gcc to convert their "``real_t``" into an ``mpfr_ptr`` then back again after calling the function.  :-)

kinke replied:
> these functions are implemented by the MPFR library

Thanks, that's what I already expected to be provided for by LLVM. We'll probably use that library too then.

ibuclaw commented on src/root/real_t.d:
```diff
@@ -0,0 +1,119 @@
+{
+    static __gshared real_t snan;
+
+    static this()
```
At what point is this expected to be executed?  (Hint, GDC replaces the ``PortInitializer`` construct with ``Port::init`` because this information can only be initialized after the backend has started).

kinke replied:
Before accessing `snan` for the first time. No idea when that is though. ;)
We won't use this struct anyway in GDC and LDC, it's only dmd's default implementation.

ibuclaw commented:
Apart from blatant namespace abuse, there's nothing surprising here.

kinke commented:
I don't think the `Darwin_64_32` autotester failures since the last commit are really caused by this PR. The `runnable/cppa.d:252` assert doesn't have anything to do with floating point, it's a C++ interop issue. What I did change in the last commit is using core.stdc.math.tan**l** instead of the double-version in `TargetReal.tan()`...

> blatant namespace abuse

No more, as the `real_t` helper functions are defined in the `TargetReal` struct now.

kinke committed:
Some more preparations for a custom real_t type

Prepare for non-implicit type conversions (e.g., from integers) and
no default-constructibility.

Based on LDC experiments with a custom real_t implementation.

ibuclaw commented on src/mtype.d:
```diff
@@ -3678,17 +3675,17 @@ public:
             case Timaginary32:
             case Tfloat32:
-                fvalue = FLT_EPSILON;
+                fvalue = real_t(FLT_EPSILON);
```
This is crying for ``float_t``.

kinke replied:
Is there ever a case where `float_t != float`, same for `double_t`?
I'd agree that using `float[_t].epsilon` would be nicer than keep on using these original C99 macros from `core.stdc.float_`.

ibuclaw replied:
Never (even GCC finally dropped ``-fshort-double`` at long last!).  However the problem is converting native -> emulation vs. just asking the emulation layer directly.

ibuclaw replied:
Also, I'm probably not making myself clear enough that GDC has a blanket ban on *all* floating point types and literals.  That includes using values from ``math.h`` that are known to the GCC internals at compile-time.  I'm not going to relax this bad either just because the front-end switched to D.

ibuclaw commented on src/mtype.d:
```diff
@@ -3678,17 +3675,17 @@ public:
             case Timaginary64:
             case Tfloat64:
-                fvalue = DBL_EPSILON;
+                fvalue = real_t(DBL_EPSILON);
```
Likewise, ``double_t``

ibuclaw commented:
Now that I've come back after a small break, I think I've come to the conclusion that this PR is all upside-down.

There are two problems to address here.

1. The need for an agnostic way to retrieve target ``float``, ``double``, and ``real`` properties (it may be different from the host for cross-compilers, *especially* for ``real``).
2. The need to support alternative soft-emulation of *all* float operations in the front-end (where the goal is for consistency rather than emulating the target).

For [1], there is replacing the macros with ``real_t.XXX``.  This seems obvious enough, but I feel that it is a faux pas.  The ``real_t`` type itself I feel should only be dealing with solving [2], as its underlying self could be anything, such as a ``BitArray`` large enough to hold 160-bits of the significand.  Of course, you wouldn't want ``real_t.mant_dig`` to return ``160``, but you *do* on occasion want to request a view of what value is being held by the type (eg: ``toFloat``, ``toDouble``, ``toReal``).

I think I've come to the conclusion (for now, or until someone convinces me otherwise ;-) that all float, double, and real properties are the best candidates for the proposed ``TargetReal`` struct, and nothing else.

For [2], there are two parts to this.  The first is using operator overloading in ``real_t`` to mimic all unary and binary operations supported by native floats, and there is introducing ``TargetReal`` to do operations that are done with some kind of function call.

Skipping the former (the ability to do ``real_t a = b + c;`` is kind of the point of the type) to discuss the latter.  I strongly feel that this name is misleading, even if in the best intentions.  As previously said, doing operations such as ``sin`` or ``tan`` has nothing in common with the target platform, it's simply a means of being able to do the intrinsic math operations supported by DMD's CTFE (GDC has a larger plethora of supported intrinsics, but I digress...) with a custom float type.  The precision is not important, so long as there's not a lack of - as above, we could return a value with higher accuracy than what *both* host and target can represent.  Neither is accuracy with runtime, although *of course* we want the end result to be correct, however see #5191 as an example of why CTFE can differ.

Emulating the target at compile-time is not the intention, so what's the point of calling it ``Target``?  Considering everything that resides there, it would be better calling it ``Math``, at least that would be an accurate description of it.

Geod24 commented:
> Emulating the target at compile-time is not the intention, so what's the point of calling it Target? Considering everything that resides there, it would be better calling it Math, at least that would be an accurate description of it.

Just to be clear: Does that mean FP operations done using CTFE will use the host's precision, and thus give different result than RT ones when cross compiling ?

ibuclaw commented:
> Just to be clear: Does that mean FP operations done using CTFE will use the host's precision, and thus give different result than RT ones when cross compiling ?

Talking specifically about GDC.
- All FP operations are done at a higher precision than *both* host and target.
- You are guaranteed to get the same *internal* result regardless of host/target combination.
- The value will be *truncated* to the precision of ``float``, ``double``, or (target)``real`` if CTFE requests it.
- The interval at which CTFE truncates precision is *different* from that of runtime.

kinke commented:
> Emulating the target at compile-time is not the intention, so what's the point of calling it Target?

Oh I finally understand your view now. Well I thought it was **all** about emulating the target at compile-time, mimicking a native compiler!

ibuclaw commented:
> Oh I finally understand your view now.

Yay!

kinke commented:
@ibuclaw: Alright, so I see the following required steps:

1. Rename `real_t.d` to `float.d`
2. Keep on defining a `real_t` type representing an arbitrary-precision floating-point type used by the front-end
3. Define a `struct FPTypeProperties { init, nan, max, min_normal, dig, ... }` and export 3 globals of this struct (`targetFloatProperties, targetDoubleProperties, targetRealProperties`) in `ddmd.root.float` and use them in `ddmd.mtype`
4. Rename `struct TargetReal` to `CTFloat` (compile-time floating-point helper, hosting static math, comparison and parse/print functions)

Did I forget something/any proposals for better names?

ibuclaw commented:
This is what I have in my "real_properties" struct.  There's an absence of nan or init, as they are generated in Port currently.
```
  longdouble maxval, minval, epsilonval;
  int64_t dig, mant_dig;
  int64_t max_10_exp, min_10_exp;
  int64_t max_exp, min_exp
```

kinke committed:
Change semantics of real_t

From a type representing a target `real` to a type representing a
compile-time `real`.

kinke commented:
@ibuclaw Reworked. GDC and LDC would have to re-define `ddmd.core.port.TargetRealProperties` based on the target and use a custom `ddmd.core.ctfloat` module for arbitrary-precision compile-time math.

ibuclaw commented:
I'm still kind of torn by putting all math related functions in it's own namespace.

One the one hand there is ``Port``, which I have no intention of using the header directly from upstream, but keep it a local copy source compatible by chucking out 90% of everything, and changing the signatures of what's remaining to satisfy the types used by gdc (which may differ from dmd).

Now you are proposing a new type called ``CTFloat``, which I also have no intention of using and will have to do something similar about it too.

From where I'm standing, it just looks cleaner to have all eggs in on basket here.

kinke commented:
Oh please, not back to that argument again! There's no more math/`real_t` stuff in `Port`, period; that's all in `CTFloat` now. Why? Cause it all depends on how `real_t` is defined.
From the LDC side, I know that we'll only change `port.d` wrt. `TargetRealProperties`, the rest will be identical to upstream. Of course there'll be a complete replacement of `ctfloat.d`, but it will include a `CTFloat` helper struct just like upstream, in order not to have to touch any of the other files.

kinke committed:
Move TargetXXXProperties template from ddmd.root.port to ddmd.target

ibuclaw commented:
After this change, there will only be ``memicmp``, ``isFloat32LiteralOutOfRange`` and ``isFloat64LiteralOutOfRange`` that are actually used by the frontend sources imported by GDC.

ibuclaw commented:
> Oh please, not back to that argument again!

Yeah, I know... :cry:  I'm trying so hard to square this circle though, just to see how it will all piece together on my side.  But I am having an internal struggle to get my head wrapped around it.

> There's no more math/real_t stuff in Port, period; that's all in CTFloat now. Why? Cause it all depends on how real_t is defined.

I could probably get on by with ``alias CTFloat = real_t``.  I'll have to dwell on it though.

kinke commented:
So you won't have to touch `port.d`, right?

For LDC, `real_t` will be a wrapper around a LLVM type, implemented in C++. It's a bit cumbersome as the LLVM type uses dynamic memory allocations (dtor, copy ctor, assignment operator etc.).
`CTFloat` will mostly represent the bridge between this `real_t` and the MPFR library used for the built-in math functionality in arbitrary precision.
Judging by your concerns, you'll probably have it all in your `real_t` type already and don't want to write forwarders in a `CTFloat` struct. The 'problem' is that DMD's `real_t` is a simple `real` which doesn't offer sin/cos/tan/sqrt/fabs/parse/sprint etc. And for LDC, I think it's quite nice to have `real_t` and the built-in math bindings separated in 2 types.

See https://github.com/ldc-developers/ldc/pull/1317 for an outline of how it could look like (no MPFR bindings yet).

kinke commented:
@ibuclaw: Did you get a chance to make up your mind yet? It's becoming more and more urgent for LDC.

ibuclaw commented:
@kinke, I've been in holidays. Maybe we should talk over dconf?  My only concern is having to break a design that already works in GDC for no apparent gain.

kinke commented:
@ibuclaw: No worries, I hardly find any time atm anyway. I won't be at dconf. :/
I've rebased the PR ~12 hours ago, and there's already the next merge conflict - damn.

kinke commented:
Rebased once more. Any rebasing operation is risky in a sense that I'll miss new `real` occurrences and that I may introduce formatting inconsistencies etc. The longer this has to wait, the less I feel comfortable with the changes without crawling through the code again.

kinke commented:
Rebased.

ibuclaw commented:
Apart from the bikeshed, there's nothing glaringly wrong here.  So let's move forward with this.

I've just thought of one feature request to add to this, though I can raise it after merge if you'd prefer.  Comment incoming...

ibuclaw commented on src/expression.d:
```diff
@@ -14695,7 +14694,7 @@ extern (C++) final class PowExp : BinExp
         e = new ScopeExp(loc, mmath);
 
-        if (e2.op == TOKfloat64 && e2.toReal() == 0.5)
+        if (e2.op == TOKfloat64 && e2.toReal() == real_t(0.5))
```
Scattered around the place are this, `real_t(1)`, `real_t(-1)`, and `real_t(0)`.  As these are so common shouldn't they instead be properties of `CTFloat`?  E.g: `CTFloat.zero`, `CTFloat.one`, `CTFloat.minusone`, `CTFloat.half`.

Having it in a cached constant would be better than having to support converting from native double or int to real_t.  Bonus motivation for GDC, all these numbers are constants in GCC's real implementation.  :-)

ibuclaw commented:
Actually, I'll raise a separate PR for that.  Let's close this.

kinke commented:
Oh wow, thanks Iain!

UplinkCoder commented:
I like this change. Thnaks @kinke
